mgr/src/server/plugins/volume_expand.cr

```diff
@@ -0,0 +1,349 @@
+        free_port = Datastore.free_port(pool.not_nil!.id, storage_unit.node.id)
+        storage_unit.port = free_port unless free_port.nil?
+      end
+
+      # No free port found
+      if storage_unit.port == 0
+        # Using halt directly from here will not work. Since it adds `next`
+        # and it just breaks the Storage_unit loop
+        no_port_available = true
+        no_port_storage_node = storage_unit.node.name
+        break
+      end
+
+      Datastore.reserve_port(pool.not_nil!.id, storage_unit.node.id, storage_unit.port)
+    end
+
+    break if no_port_available
+  end
+
+  if no_port_available
+    halt(env, status_code: 400, response: ({"error": "No free Port available in #{no_port_storage_node}"}).to_json)
+  end
+
+  # Generate Services and Volfiles if Volume to be started
+  services, volfiles = services_and_volfiles(req)
+
+  action = ACTION_VOLUME_CREATE
+  req.state = volume.state
+  if volume.state != "Started"
+    action = ACTION_VOLUME_CREATE_STOPPED
+  end
+
+  # Volume create action {req, services, volfiles}
+  resp = dispatch_action(
+    action,
+    pool_name,
+    nodes,
+    {services, volfiles, req}.to_json
+  )
+
+  if !resp.ok
+    halt(env, status_code: 400, response: node_errors("Failed to expand Volume", resp.node_responses).to_json)
+  end
+
+  # # Move to end, after fix-layout
+  # # Save Services details... Causing exception of unique constraint failed
+  # services.each do |node_id, svcs|
+  #   svcs.each do |svc|
+  #     # Enable each Services
+  #     begin
+  #       puts "in enable service"
+  #       Datastore.enable_service(pool.not_nil!.id, node_id, svc)
+  #     rescue ex : Exception
+  #       # Avoid adding same service into DB
+
+  #       puts "in execption"
+
+  #       # update servcice in DB
+  #       # Write a node action to call svc.restart
+
+  #       # puts "after restarting shd"
+
+  #       Datastore.update_service(pool.not_nil!.id, node_id, svc)
+
+  #       puts "after updating service in DB"
+  #     end
+  #   end
+  # end
+
+  storage_units = Hash(String, Hash(String, MoanaTypes::StorageUnit)).new
+  resp.node_responses.each do |node, node_resp|
+    storage_units[node] = Hash(String, MoanaTypes::StorageUnit).from_json(node_resp.response)
+  end
+
+  # Save Ports details and Update Storage unit metrics and FS type
+  req.distribute_groups.each do |dist_grp|
+    dist_grp.storage_units.each do |storage_unit|
+      storage_unit.metrics = storage_units[storage_unit.node.id][storage_unit.path].metrics
+      storage_unit.fs = storage_units[storage_unit.node.id][storage_unit.path].fs
+    end
+  end
+
+  set_volume_metrics(req)
+
+  puts "type", req.type
+  puts "last expand"
+  puts req
+  puts req.to_json
+
+  # Save Volume info
+  Datastore.update_volume(pool.not_nil!.id, req, volume.not_nil!.distribute_groups.size)
+
+  puts "sending restart req"
+
+  resp = dispatch_action(
+    ACTION_RESTART_SHD_SERVICE,
+    pool_name,
+    nodes,
```

Comment:
This should be all nodes from Volume, not just the new nodes.

Reply:
`nodes = participating_nodes(pool_name, new_volume)`

Reply:
appended new and existing nodes. done